File: bidi-stream/NovaS2SWebSocketExample/app/src/main/java/org/example/utility/InputEventsInteractObserver.java
package org.example.utility;

import static com.google.common.base.Preconditions.checkNotNull;

import org.reactivestreams.Subscriber;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.services.bedrockruntime.model.BidiInputPayloadPart;
import software.amazon.awssdk.services.bedrockruntime.model.InvokeModelWithBidiStreamInput;

public class InputEventsInteractObserver implements InteractObserver<String> {
    private static final Logger log = LoggerFactory.getLogger(InputEventsInteractObserver.class);
    private static final String SESSION_END = """
                {
                    "event": {
                        "sessionEnd": {}
                    }
                }""";
    private final Subscriber<InvokeModelWithBidiStreamInput> subscriber;

    public InputEventsInteractObserver(Subscriber<InvokeModelWithBidiStreamInput> publisher) {
        this.subscriber = checkNotNull(publisher, "subscriber cannot be null");
    }

    @Override
    public void onNext(String msg) {
        log.info("publishing message {}", msg);
        this.subscriber.onNext(inputBuilder(msg));
    }

    @Override
    public void onComplete() {
        this.subscriber.onNext(inputBuilder(SESSION_END));
        this.subscriber.onComplete();
    }

    @Override
    public void onError(Exception error) {
       new RuntimeException(error.getMessage());
    }

    private BidiInputPayloadPart inputBuilder (String input) {
        return InvokeModelWithBidiStreamInput.chunkBuilder()
                .bytes(SdkBytes.fromUtf8String(input))
                .build();
    }
}

File: speech-to-speech/amazon-nova-2-sonic/sample-codes/websocket-dotnet/Utility/InputEventsInteractObserver.cs
using System.Collections.Concurrent;
using System.Text;
using Amazon.BedrockRuntime.Model;
using Microsoft.Extensions.Logging;

namespace NovaSonicWebSocket.Utility;

public class InputEventsInteractObserver : IInteractObserver<string>
{
    private static readonly string SESSION_END = """
        {
            "event": {
                "sessionEnd": {}
            }
        }
        """;

    private readonly BlockingCollection<IInvokeModelWithBidirectionalStreamInputEvent> _messageQueue;
    private readonly ILogger _logger;

    public InputEventsInteractObserver(BlockingCollection<IInvokeModelWithBidirectionalStreamInputEvent> messageQueue, ILogger logger)
    {
        _messageQueue = messageQueue ?? throw new ArgumentNullException(nameof(messageQueue));
        _logger = logger;
    }

    public void OnNext(string msg)
    {
        //_logger.LogInformation("Publishing message {Message}", msg);
        _messageQueue.Add(CreateInputEvent(msg));
    }

    public void OnComplete()
    {
        _messageQueue.Add(CreateInputEvent(SESSION_END));
        _messageQueue.CompleteAdding();

    }

    public void OnError(Exception error)
    {
        _messageQueue.CompleteAdding();
    }

    private BidirectionalInputPayloadPart CreateInputEvent(string input)
    {
        return new BidirectionalInputPayloadPart
        {
            Bytes = new MemoryStream(Encoding.UTF8.GetBytes(input))
        };
    }
}
